Brokerages/Zerodha/ZerodhaBrokerage.cs

```diff
@@ -0,0 +1,1352 @@
+
+                if (historyRequest.StartTimeUtc >= historyRequest.EndTimeUtc)
+                {
+                    throw new ArgumentException("Invalid date range specified");
+                }
+
+                var start = historyRequest.StartTimeUtc.ConvertTo(DateTimeZone.Utc, TimeZones.Kolkata);
+                var end = historyRequest.EndTimeUtc.ConvertTo(DateTimeZone.Utc, TimeZones.Kolkata);
+
+                var history = Enumerable.Empty<BaseData>();
+
+                if (Config.GetBool("zerodha-history-subscription"))
+                {
+                    switch (historyRequest.Resolution)
+                    {
+                        case Resolution.Minute:
+                            history = GetHistoryForPeriod(historyRequest.Symbol, start, end, historyRequest.Resolution, "minute");
+                            break;
+
+                        case Resolution.Hour:
+                            history = GetHistoryForPeriod(historyRequest.Symbol, start, end, historyRequest.Resolution, "60minute");
+                            break;
+
+                        case Resolution.Daily:
+                            history = GetHistoryForPeriod(historyRequest.Symbol, start, end, historyRequest.Resolution, "day");
+                            break;
+                    }
+                }
+
+                foreach (var baseData in history)
+                {
+                    yield return baseData;
+                }
+            }
+        }
+
+        private IEnumerable<BaseData> GetHistoryForPeriod(Symbol symbol, DateTime start, DateTime end, Resolution resolution, string zerodhaResolution)
+        {
+            Log.Debug("ZerodhaBrokerage.GetHistoryForPeriod();");
+            var scripSymbol = _symbolMapper.GetZerodhaInstrumentToken(symbol.Value, symbol.ID.Market);
+            var candles = _kite.GetHistoricalData(scripSymbol.ToStringInvariant(), start, end, zerodhaResolution);
+
+            if (!candles.Any())
+            {
+                OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Warning, "NoHistoricalData",
+                    $"Exchange returned no data for {symbol} on history request " +
+                    $"from {start:s} to {end:s}"));
+            }
+
+            var period = resolution.ToTimeSpan();
+
+            foreach (var candle in candles)
+            {
+                yield return new TradeBar()
+                {
+                    Time = candle.TimeStamp.ConvertFromUtc(TimeZones.Kolkata),
+                    Symbol = symbol,
+                    Low = candle.Low,
+                    High = candle.High,
+                    Open = candle.Open,
+                    Close = candle.Close,
+                    Volume = candle.Volume,
+                    Value = candle.Close,
+                    DataType = MarketDataType.TradeBar,
+                    Period = period,
+                    EndTime = candle.TimeStamp.Add(period).ConvertFromUtc(TimeZones.Kolkata)
```

Comment:
EndTime is created automatically using the period

Reply:
Careful, the default value for `Period` is 1 minute, `EndTime` is created automatically based on `Time` and `Period` see https://github.com/QuantConnect/Lean/blob/master/Common/Data/Market/TradeBar.cs#L109 . Suggest using the `TradeBar` constructors to avoid these issues, it's error-prone and has happened before 😄